Modules/_posixsubprocess.c

```diff
@@ -677,6 +705,51 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
         cwd_obj2 = NULL;
     }
 
+    if (groups_list != Py_None) {
+        Py_ssize_t i;
+        int gid;
+
+        if (!PyList_Check(groups_list)) {
+            PyErr_SetString(PyExc_TypeError, "setgroups argument must be a list");
+            goto cleanup;
+        }
+        num_groups = PySequence_Size(groups_list);
+
+        if (num_groups < 0)
+            goto cleanup;
+
+        if (num_groups > MAX_GROUPS) {
+            PyErr_SetString(PyExc_ValueError, "too many groups");
+            goto cleanup;
+        }
+
+        for(i = 0; i < num_groups; i++) {
+            PyObject *elem;
+            elem = PySequence_GetItem(groups_list, i);
+            if (!elem)
+                goto cleanup;
+            if (!PyLong_Check(elem)) {
+                PyErr_SetString(PyExc_TypeError,
+                                "groups must be integers");
+                Py_DECREF(elem);
+                goto cleanup;
+            } else {
+                /* In posixmodule.c UnsignedLong is used as a fallback value
+                 * if the value provided does not fit in a Long. Since we are
+                 * already doing the bounds checking on the Python side, we
+                 * can go directly to an UnsignedLong here. */
+                gid = PyLong_AsUnsignedLong(elem);
```

Comment:
The type of `gid` is `int`, so truncation is possible in this assignment.

Reply:
OK, I can make gid be `unsigned long` and check if it is greater than INT_MAX before assigning it to `gid_t`